miniapps/blocksolvers/div_free_solver.hpp

```diff
@@ -0,0 +1,252 @@
+   void DataFinalize();
+public:
+   DFSDataCollector(int order, int num_refine, ParMesh *mesh,
+                    const Array<int>& ess_attr, const DFSParameters& param);
+
+   void CollectData();
+
+   const DFSData& GetData() const { return data_; }
+
+   unique_ptr<ParFiniteElementSpace> hdiv_fes_;
+   unique_ptr<ParFiniteElementSpace> l2_fes_;
+   unique_ptr<ParFiniteElementSpace> hcurl_fes_;
+};
+
+/// Abstract solver class for Darcy's flow
+class DarcySolver : public Solver
+{
+protected:
+   Array<int> offsets_;
+public:
+   DarcySolver(int size0, int size1) : Solver(size0 + size1), offsets_(3)
+   { offsets_[0] = 0; offsets_[1] = size0; offsets_[2] = height; }
+   virtual int GetNumIterations() const = 0;
+};
+
+/// Solver for B * B^T
+class BBTSolver : public Solver
+{
+   OperatorPtr BBT_;
+   OperatorPtr BBT_prec_;
+   CGSolver BBT_solver_;
+   bool B_has_nullity_one_;
+public:
+   BBTSolver(const HypreParMatrix &B, bool B_has_nullity_one=false,
+             IterSolveParameters param=IterSolveParameters());
+   virtual void Mult(const Vector &x, Vector &y) const;
+   virtual void SetOperator(const Operator &op) { }
+};
+
+/// Block diagonal solver for A, each block is inverted by direct solver
+class BlockDiagSolver : public Solver
+{
+   mutable SparseMatrix block_dof_;
+   mutable Array<int> local_dofs_;
+   mutable Vector sub_rhs_;
+   mutable Vector sub_sol_;
+   Array<DenseMatrixInverse> block_solver_;
+public:
+   BlockDiagSolver(const OperatorPtr& A, SparseMatrix block_dof);
+   virtual void Mult(const Vector &x, Vector &y) const;
+   virtual void MultTranspose(const Vector &x, Vector &y) const { Mult(x, y); }
+   virtual void SetOperator(const Operator &op) { }
+};
+
+/// Solver for local problems in SchwarzSmoother
+class LocalSolver : public Solver
+{
+   DenseMatrix local_system_;
+   DenseMatrixInverse local_solver_;
+   const int offset_;
+public:
+   LocalSolver(const DenseMatrix &M, const DenseMatrix &B);
+   virtual void Mult(const Vector &x, Vector &y) const;
+   virtual void SetOperator(const Operator &op) { }
+};
+
+/// non-overlapping additive Schwarz for saddle point problems
+class SchwarzSmoother : public Solver
+{
+   const SparseMatrix& agg_hdivdof_;
+   const SparseMatrix& agg_l2dof_;
+   OperatorPtr coarse_l2_projector_;
+
+   Array<int> offsets_;
+   mutable Array<int> offsets_loc_;
+   mutable Array<int> hdivdofs_loc_;
+   mutable Array<int> l2dofs_loc_;
+   Array<OperatorPtr> solvers_loc_;
+public:
+   SchwarzSmoother(const HypreParMatrix& M,
+                   const HypreParMatrix& B,
+                   const SparseMatrix& agg_hdivdof,
+                   const SparseMatrix& agg_l2dof,
+                   const HypreParMatrix& P_l2,
+                   const HypreParMatrix& Q_l2);
+   virtual void Mult(const Vector & x, Vector & y) const;
+   virtual void SetOperator(const Operator &op) { }
+};
+
+/// Relaxation on the kernel of divergence
+class KernelSmoother : public Solver
```

Comment:
This might be useful in the MFEM library. In fact it has nothing to do with divergence, I think you could use exactly this object for a Hiptmair smoother for H(curl) problems, for example.